### PR TITLE
feat: add more components to the wasm-pack compatible list

### PR DESCRIPTION
**Pull request by waynexia**
## Which issue does this PR close?

Related to https://github.com/apache/arrow-datafusion/pull/7633

## Rationale for this change

Make more components to be `wasm-unknown-unknown` compatible.

About the background and motivation: I'm trying to continue my side project that builds a web playground based on wasm for datafuson https://github.com/waynexia/datafusion-playground

## What changes are included in this PR?

Add supports for `datafusion-physical-plan`, `datafusion-execution` and `datafusion`.

`datafusion-physical-plan` and `datafusion-execution` doesn't contain any logical changes. It works after I remove some unnecessary feature gates or dependence.

Some methods and logic in `datafusion` are commented out. The affected file is `datafusion/core/src/datasource/listing/url.rs`, all about parsing URL. This is a limitation from our dependence `url`. After some shallow investigates, I guess it's caused by `OsStr` from std isn't available in arch `wasm-unknown-unknown`.

## Are these changes tested?

Yes

## Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/substrait/Cargo.toml [hunk 1]
```diff
@@ -36,6 +36,8 @@ object_store = { workspace = true }
 prost = "0.12"
 prost-types = "0.12"
 substrait = "0.22.1"
+
+[dev-dependencies]
 tokio = "1.17"
```
Comment (waynexia) [inline datafusion/substrait/Cargo.toml:41]: I move this to dev-dependencies casually. But `datafusion-substrait` and `datafusion-proto` are not supported yet.

File: datafusion/wasmtest/Cargo.toml [hunk 2]
```diff
@@ -37,11 +37,14 @@ crate-type = ["cdylib", "rlib",]
 # all the `std::fmt` and `std::panicking` infrastructure, so isn't great for
 # code size when deploying.
 console_error_panic_hook = { version = "0.1.1", optional = true }
+datafusion = { path = "../core", default-features = false }
 
 datafusion-common = { workspace = true }
+datafusion-execution = { workspace = true }
```
Comment (alamb) [inline datafusion/wasmtest/Cargo.toml:43]: I double checked and this addition means that the wasm ci job now checks these crates
https://github.com/apache/arrow-datafusion/blob/8353a2ca2fd1f0ed5fc764b7463dfbcaa033ceef/.github/workflows/rust.yml#L192-L207

Here is the example CI run from this PR showing the new crates building in wasm https://github.com/apache/arrow-datafusion/actions/runs/7504130188/job/20430447784?pr=8843#step:6:194

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @waynexia

**Comment by waynexia** [comment]
Thanks for your review @alamb

**Pull request merged**